### PR TITLE
TGUI input box framework

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Creates the _framework_ for two new TGUI input boxes that can be toggled via game prefs.
This does not convert any actual inputs to TGUI
This does not convert any tgui_list_inputs into being toggleable

### Example pictures
Input on a hand labeler. This has a MAX_LENGTH set, so it can be invalidated. Cancel always returns null. Enter button submits, if valid.
text input
(OUTDATED) Multiline input on newscaster. Newer version fills the window with a section, like the others
multiline
Sheets from a stack
number input

## Why It's Good For The Game

So I did...
- Much sleeker input boxes
- Result should be a in place swap for most occurrences of input
- Renders casting as text/num/null obsolete but still doable
- Input validation from both sides handled
- Prefs toggle if you don't like the look

## Changelog

:cl:
add: TGUI input boxes are on the way! You can find new preferences in the menu. They will be on by default.
/:cl:

### CODE REVIEW

**Comment by Watermelon914** [comment]
post a screenshot if you can

**Comment by stylemistake** [comment]
Couldn't you just reuse the tgui_input? With perhaps a flag that makes it a multiline input?

**Comment by jlsnow301** [comment]
> Couldn't you just reuse tgui_input? With perhaps a flag that makes it a multiline input?

There is only input_list and alert

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
-  name: "Lock tgui to main monitor",
-  category: "UI",
-  description: "Locks tgui windows to your main monitor.",
+  name: 'Lock tgui to main monitor',
+  category: 'UI',
+  description: 'Locks tgui windows to your main monitor.',
+  component: CheckboxInput,
+};
+
+export const tgui_input: FeatureToggle = {
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx:17]: ABC order

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
+  name: 'Enable tgui input boxes',
+  category: 'UI',
+  description: 'Renders input boxes in TGUI.',
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx:20]:
```suggestion
  description: 'Renders input boxes in tgui.',
```

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx [hunk 3]
```diff
@@ -1,15 +1,22 @@
-import { CheckboxInput, FeatureToggle } from "../base";
+import { CheckboxInput, FeatureToggle } from '../base';
 
 export const tgui_fancy: FeatureToggle = {
-  name: "Enable fancy tgui",
-  category: "UI",
-  description: "Makes tgui windows look better, at the cost of compatibility.",
+  name: 'Enable fancy tgui',
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx:4]: eh whatever

**Comment by Mothblocks** [comment]
Those Submit/cancel buttons are fairly far from each other, can you make them more like the current ones?

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx (hunk 3 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/tgui.tsx:4]: sorry, just prettier doing its thing - i think @stylemistake set the single quotation flag

**Comment by jlsnow301** [comment]
> Those Submit/cancel buttons are fairly far from each other, can you make them more like the current ones?

Yes

**Comment by jlsnow301** [comment]
Considering removing the nanotrasen background && swapping the submit / cancel buttons

**Comment by Watermelon914** [comment]
I think there should be another section covering the input box and the buttons, I find it bizarre to look at otherwise

**Comment by nfreader** [comment]
> swapping the submit / cancel buttons

> This change follows the usability principle of placing positive user actions on the right to convey a choice that moves the user forward in the flow, and negative actions on the left, an action that moves the user backward in the flow.   

From: https://help.duo.com/s/article/7231?language=en_US

**Comment by jlsnow301** [comment]
> > swapping the submit / cancel buttons
> 
> 
> 
> > This change follows the usability principle of placing positive user actions on the right to convey a choice that moves the user forward in the flow, and negative actions on the left, an action that moves the user backward in the flow.   
> 
> 
> 
> From: https://help.duo.com/s/article/7231?language=en_US

As I was discussing in the disc when this was brought up: I see the point and agree; the problem is user's muscle memory on an ancient game and some other dialogue windows cannot be changed, so you'll have both left-right and right-left in game which will end up confusing-ish

**Comment by jlsnow301** [comment]
> I think there should be another section covering the input box and the buttons, I find it bizarre to look at otherwise
here are some examples with sections a) over input box & buttons, b) default, c) over the entire box

**Comment by Watermelon914** [comment]
> > I think there should be another section covering the input box and the buttons, I find it bizarre to look at otherwise
> 
> here are some examples with sections a) over input box & buttons, b) default, c) over the entire box 

What if you had it as a section, and then the label text set as the section title?

**Comment by jlsnow301** [comment]
> What if you had it as a full section, and then the label text set as the section title? Also, consider making the buttons fluid so that they fill all the empty space

I don't think this will look right in cases where the message text is two or more lines. i will do it later, and show some photos of it.
i like the buttons as is - i'm going for the direct replacement and they match the sizing of the original.
since i'm already making a preference, it wouldn't pain me to make more preferences for button sizing and arrangement

File: tgui/packages/tgui/interfaces/NumberInput.tsx [hunk 4]
```diff
@@ -41,45 +49,37 @@ export const NumberInput = (_, context) => {
     setInputIsValid(validateInput(value, max_value, min_value));
     setInput(value);
   };
-  // Dynamically changes the window height based on the message.
-  const windowHeight = 130 + Math.ceil(message.length / 5);
+  /** Dynamically changes the window height based on the message. */
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NumberInput.tsx]: Don't change comments into doc blocks, this is not documentation.

File: tgui/packages/tgui/interfaces/NumberInput.tsx [hunk 5]
```diff
@@ -30,9 +36,11 @@ export const NumberInput = (_, context) => {
   const onType = (event) => {
     event.preventDefault();
     let value = event.target.value;
-    /** Cuts leading zeros */
-    if (!isNaN(value)) {
-      value = value * 1;
+    /** Checks the input for leading 0s and removes them.
+     *  The if check disables submit on window load.
+     */
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NumberInput.tsx]: Convert into //

File: tgui/packages/tgui/interfaces/TextInput.tsx [hunk 6]
```diff
@@ -33,43 +39,34 @@ export const TextInput = (_, context) => {
     setInputIsValid(validateInput(target.value, max_length));
     setInput(target.value);
   };
-  // Dynamically changes the window height based on the message.
+  /** Dynamically changes the window height based on the message. */
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/TextInput.tsx]: Same

File: tgui/packages/tgui/interfaces/common/InputButtons.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+export type Preferences = {
+  large_buttons: boolean;
+  swapped_buttons: boolean;
+};
+
+/** The buttons shown at bottom. Will display the error
+ * if the input is invalid.
+ */
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/InputButtons.tsx:25]: Same, this is not documentation, and don't write on the top part `/**` if you do.

**Review by stylemistake — APPROVED** [review]

**Pull request merged**